### PR TITLE
New Adapter: Nativo

### PR DESCRIPTION
**Pull request by rafataveira**

### CODE REVIEW

**Comment by rafataveira** [comment]
@onkarvhanumante Is there any ETA for reviewing this PR?
When is the next version expected to be released?
TIA

**Comment by onkarvhanumante** [comment]
@rafataveira  PR tests are failing due to file miss format. Requesting to format file


```
Run ./validate.sh --nofmt --cov --race 10
gofmt needs to be run, 2 files have issues.  Below is a list of files to review:
adapters/nativo/nativo.go
adapters/nativo/nativo_test.go
```

**Comment by ashishshinde-pubm** [comment]
@rafataveira Does this bidder supports `request.test` parameter and respond with dummy response ? or is there any way this bidder can return dummy response for test purpose ?

File: static/bidder-info/nativo.yaml
Comment (przemkaczmarek) [inline static/bidder-info/nativo.yaml:3]: I've sent an email to this address to confirm it is correct. Please respond to the email with a "received" message.

File: static/bidder-params/nativo.json
Comment (przemkaczmarek) [inline static/bidder-params/nativo.json:6]: NIP, please delete white line

File: adapters/nativo/nativo_test.go
Comment (przemkaczmarek) [inline adapters/nativo/nativo_test.go:13]: why is it "https://foo.io/?src=prebid" ? 
It should be endpoint from yaml

File: static/bidder-info/nativo.yaml
Comment (przemkaczmarek) [inline static/bidder-info/nativo.yaml:1]: "esi?ntv_epid=7" does not look fine

File: adapters/nativo/nativo.go
Comment (onkarvhanumante) [inline adapters/nativo/nativo.go:67]: would recommend using adapter utils as done in Appnexus adapterhttps://github.com/prebid/prebid-
server/blob/3295c4119324f6d33d1d9851872cd9b2e594f3f3/adapters/appnexus/appnexus.go#L164-L176

File: adapters/nativo/nativo.go
Comment (onkarvhanumante) [inline adapters/nativo/nativo.go:101]: Prebid server expects the media type to be explicitly set in the adapter response. Therefore, we strongly recommend implementing a pattern where the adapter server sets the [MType](https://github.com/prebid/openrtb/blob/main/openrtb2/bid.go#L334) field in the response to accurately determine the media type for the impression.

File: static/bidder-info/nativo.yaml
Comment (onkarvhanumante) [inline static/bidder-info/nativo.yaml:4]: verified gvl id

```
% curl https://vendor-list.consensu.org/v3/vendor-list.json | jq '.vendors."263"'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  626k  100  626k    0     0  5203k      0 --:--:-- --:--:-- --:--:-- 5218k
{
  "id": 263,
  "name": "Nativo, Inc.",
  "purposes": [
    1,
    3,
    4
  ],
  "legIntPurposes": [
    2,
    7,
    9,
    10
  ],
  "flexiblePurposes": [],
  "specialPurposes": [
    1,
    2,
    3
  ],
  "features": [
    2
  ],
  "specialFeatures": [],
  "cookieMaxAgeSeconds": 31536000,
  "usesCookies": true,
  "cookieRefresh": true,
  "usesNonCookieAccess": true,
  "dataRetention": {
    "stdRetention": 730,
    "purposes": {
      "2": 0,
      "3": 45,
      "4": 45
    },
    "specialPurposes": {
      "2": 0,
      "3": 1826
    }
  },
  "urls": [
    {
      "langId": "en",
      "privacy": "https://www.nativo.com/interest-based-ads",
      "legIntClaim": "https://www.nativo.com/interest-based-ads#how-we-use-your-personal-data-and-why"
    }
  ],
  "dataDeclaration": [
    1,
    2,
    3,
    5,
    6,
    8,
    10,
    11
  ],
  "deviceStorageDisclosureUrl": "https://iab.nativo.com/tcf-disclosures.json"
}
```

File: adapters/nativo/nativo.go
Comment (onkarvhanumante) [inline adapters/nativo/nativo.go:30]: should remove extra line space on line 29

File: static/bidder-info/nativo.yaml
Comment (onkarvhanumante) [inline static/bidder-info/nativo.yaml:19]: @rafataveira  can redirect url be `https` endpoint?

Comment (onkarvhanumante) [inline static/bidder-info/nativo.yaml:19]: redirect url is not working.

```
http://jadserve.postrelease.com/suid/101787?gdpr=&gdpr_consent=&us_privacy=&ntv_gpp_consent=&ntv_r=https%3A%2F%2Fprebid.adnxs.com%2Fpbs%2Fv1%2Fsetuid%3Fbidder%3DNativo%26gdpr%3D%26gdpr_consent%3D%26f%3Di%26uid%3D%22NTV_USER_ID%22
```

In above case redirect url was not called.

File: static/bidder-info/nativo.yaml
Comment (rafataveira) [reply static/bidder-info/nativo.yaml:3]: It is correct. I've just replied to the email.

File: static/bidder-info/nativo.yaml
Comment (rafataveira) [reply static/bidder-info/nativo.yaml:1]: this is correct, that's the url to our exchange server, in this case, prebid server.

**Comment by rafataveira** [comment]
@onkarvhanumante I will review all comments and make the changes accordingly.

File: static/bidder-info/nativo.yaml
Comment (przemkaczmarek) [reply static/bidder-info/nativo.yaml:3]: mail confirmed

File: static/bidder-params/nativo.json
Comment (rafataveira) [reply static/bidder-params/nativo.json]: done

File: adapters/nativo/nativo_test.go
Comment (rafataveira) [reply adapters/nativo/nativo_test.go]: done

File: adapters/nativo/nativo.go
Comment (rafataveira) [reply adapters/nativo/nativo.go]: done

File: adapters/nativo/nativo.go
Comment (rafataveira) [reply adapters/nativo/nativo.go]: done

File: adapters/nativo/nativo.go
Comment (rafataveira) [reply adapters/nativo/nativo.go]: done

File: static/bidder-info/nativo.yaml
Comment (rafataveira) [reply static/bidder-info/nativo.yaml]: done

**Comment by bsardo** [comment]
Hi @rafataveira, your tests are failing because you need to run `go fmt`.

**Review by przemkaczmarek — APPROVED** [review]

**Comment by onkarvhanumante** [comment]
@rafataveira , redirect url is not working

@przemkaczmarek could you check if redirect url is working for you

```
http://jadserve.postrelease.com/suid/101787?gdpr=&gdpr_consent=&us_privacy=&ntv_gpp_consent=&ntv_r=https%3A%2F%2Fprebid.adnxs.com%2Fpbs%2Fv1%2Fsetuid%3Fbidder%3DNativo%26gdpr%3D%26gdpr_consent%3D%26f%3Di%26uid%3D%22NTV_USER_ID%22
```

**Comment by rafataveira** [comment]
@onkarvhanumante you need to use the secure url https:// provided in the nativo.yaml instead of http://

`https://jadserve.postrelease.com/suid/101787?gdpr=&gdpr_consent=&us_privacy=&ntv_gpp_consent=&ntv_r=https%3A%2F%2Fprebid.adnxs.com%2Fpbs%2Fv1%2Fsetuid%3Fbidder%3DNativo%26gdpr%3D%26gdpr_consent%3D%26f%3Di%26uid%3D%22NTV_USER_ID%22`

**Comment by przemkaczmarek** [comment]
works:
image

**Review by przemkaczmarek — DISMISSED** [review]

**Comment by rafataveira** [comment]
@bsardo This is the result of the redirect I get:
https://prebid.adnxs.com/pbs/v1/setuid?bidder=Nativo&gdpr=&gdpr_consent=&f=i&uid=%22c0ed3c17-4ba0-4d2e-9cf2-cbf8f9037793%22
image

**Comment by bsardo** [comment]
> @bsardo This is the result of the redirect I get: https://prebid.adnxs.com/pbs/v1/setuid?bidder=Nativo&gdpr=&gdpr_consent=&f=i&uid=%22c0ed3c17-4ba0-4d2e-9cf2-cbf8f9037793%22 image

Hi @rafataveira, I suggest testing your user sync with localhost so that your bidder is known and so that you can ensure your server redirects to the setuid endpoint of the PBS host, which in this case will be the localhost setuid endpoint. Start this process by first hitting the localhost PBS cookie sync endpoint:
`POST localhost:8000/cookie_sync` with body `{"bidders": ["nativo"]}` and that will give you the url that would be dropped as a pixel kicking off the cookie sync that should ultimately redirect to the setuid endpoint.

**Comment by rafataveira** [comment]
> > @bsardo This is the result of the redirect I get: https://prebid.adnxs.com/pbs/v1/setuid?bidder=Nativo&gdpr=&gdpr_consent=&f=i&uid=%22c0ed3c17-4ba0-4d2e-9cf2-cbf8f9037793%22 image
> 
> Hi @rafataveira, I suggest testing your user sync with localhost so that your bidder is known and so that you can ensure your server redirects to the setuid endpoint of the PBS host, which in this case will be the localhost setuid endpoint. Start this process by first hitting the localhost PBS cookie sync endpoint: `POST localhost:8000/cookie_sync` with body `{"bidders": ["nativo"]}` and that will give you the url that would be dropped as a pixel kicking off the cookie sync that should ultimately redirect to the setuid endpoint.

I found the issue with the double quotes on userMacro. Now it should work as expected.

**Comment by bsardo** [comment]
@rafataveira we're still seeing an issue with the user sync. Your server is responding with a 200. Since you've classified this user sync as type redirect, your server should be responding with a 302 redirect to the url defined in the `ntv_r` parameter with the user id macro populated.

**Comment by rafataveira** [comment]
@bsardo I couldn't reproduce this behaviour. As you can see, our server responds with 302 and correctly redirects to the URL defined in the `ntv_r` parameter with the user ID macro populated.
image
This is the same cookie sync URL used by Prebid Server Java version and other partners.

**Comment by rafataveira** [comment]
@onkarvhanumante @przemkaczmarek @bsardo, is the cookie sync URL the only thing preventing you from approving this PR? We have some clients waiting for the GO version to go live for quite some time. Could you please go ahead and approve it? The cookie sync URL is the same as prebid-server-java and other partners, and it is working as expected.

**Comment by bsardo** [comment]
> @onkarvhanumante @przemkaczmarek @bsardo, is the cookie sync URL the only thing preventing you from approving this PR? We have some clients waiting for the GO version to go live for quite some time. Could you please go ahead and approve it? The cookie sync URL is the same as prebid-server-java and other partners, and it is working as expected.

Hi @rafataveira, yes the user sync is the only thing holding this up. We tried the URL you provided as well and it is still not working for me or Peter and we are based in the US and Poland respectively in case this is a geographic region issue. We are still seeing a 200 instead of a 302 redirect. This issue needs to be resolved before merging this PR as is unless you want to break the user sync functionality out into a separate PR so that the rest of this can be merged now. At this point, I am leaving the user sync investigation up to you though I will triple check with another team member that the 200 issue is occurring and report back later today.

**Comment by rafataveira** [comment]
@bsardo I see what is happening to your validation. This is a production cookie sync URL which has different kinds of internal validations to avoid fraud and won't redirect if a possible fraud is detected. It works for me because I am in a VPN environment. This has nothing to do with our GO version adapter at all. If this is still a blocker for you to approve this PR, (even though it is not a problem) we can remove it from this PR and also, change our validation for Prebid Server to always redirect. Let me know what you recommend.

**Review by przemkaczmarek — DISMISSED** [review]

**Comment by bsardo** [comment]
@rafataveira we're required to verify the user sync endpoints works. Can you either provide details on how to construct a request so that it is not flagged by your servers as potential fraud or change your server validations so that the URL generated by the cookie sync endpoint always results in a redirect? I'll leave that decision to you at this time as I'm not aware of how your validation works.

**Comment by rafataveira** [comment]
@bsardo we have pushed a fix for the redirect validation. Can you please validate and if all looks good can we merge the PR?

**Comment by aparekh-nativo** [comment]
@bsardo can we get an update on the cookie sync validation ?

**Comment by bsardo** [comment]
Hi @aparekh-nativo, @rafataveira tests are failing and your PR branch does not run because you have some incorrect package references after merging with master. All package imports should reference v3 instead of v2. It looks like the offending files are `adapter_builders.go`, `nativo.go` and `nativo_test.go`. Yesterday we released PBS 3.0 which updated all package references to v3. See #4029.

**Comment by rafataveira** [comment]
@bsardo The merge issue and test failures are fixed. Could you please go ahead and merge it?

**Comment by bsardo** [comment]
Hi @rafataveira, now I see the redirect working but the user id macro `NTV_USER_ID` is not being populated:
Screenshot 2024-11-04 at 5 13 40 PM

**Comment by rafataveira** [comment]
@bsardo could you please recheck it? We pushed a fixed this morning to solve this issue.

File: static/bidder-info/nativo.yaml
Comment (bsardo) [inline static/bidder-info/nativo.yaml:19]: Confirmed redirect user sync is working. Observed 302 to PBS setuid endpoint with a user ID which is successfully added to cookie.
Screenshot 2024-11-08 at 9 09 40 AM

File: adapters/nativo/nativo.go
Comment (bsardo) [inline adapters/nativo/nativo.go:83]: Nitpick: I suggest just passing in what you need so instead of copying the `bid` object just pass in `bidImpID string` as the first parameter.

File: adapters/nativo/nativo.go
Comment (bsardo) [inline adapters/nativo/nativo.go:95]: Can you add a supplemental JSON test where the bid returns an imp ID that is different from what is in the request so that this code path is executed?

File: adapters/nativo/nativotest/supplemental/200-response-from-nativo.json
Comment (bsardo) [inline adapters/nativo/nativotest/supplemental/200-response-from-nativo.json:1]: You can delete this test. It is exactly the same as `204-response-from-nativo.json`.

File: adapters/nativo/nativo_test.go
Comment (bsardo) [inline adapters/nativo/nativo_test.go:13]: I suggest using a fake URL here and in all of your JSON test files for maintenance purposes.

File: adapters/nativo/nativotest/exemplary/banner-app.json
Comment (bsardo) [inline adapters/nativo/nativotest/exemplary/banner-app.json:98]: I see this field `seatbid[i].type` is in all of your JSON tests. This is not a valid ORTB bid response field so I suggest removing it as it is misleading. There is a field `seatbid[i].bid[j].mtype` which can be used to indicate the media type and this is strongly recommended instead of deriving the media type from impression IDs via your function `getMediaTypeForImp` but that is up to you if you want to make use of it at this time.

At a minimum, please remove this `"type"` field from all of your JSON test mock responses.

**Review by bsardo — COMMENTED** [review]
@rafataveira your cookie sync looks good now and the adapter looks good overall. I just took over review responsibilities from Onkar as he is currently unavailable. I left just a few more comments. Please take a look, update and ping me when ready so I can prioritize re-review. Thanks for your patience.

**Comment by rafataveira** [comment]
@bsardo all the comments have been addressed. could you please validate?

**Comment by rafataveira** [comment]
@bsardo Following up with this. Are you able to review it again and merge it? Thank you.

**Review by bsardo — APPROVED** [review]

**Comment by bsardo** [comment]
@przemkaczmarek please give this another look given recent changes.

**Review by przemkaczmarek — APPROVED** [review]

**Comment by rafataveira** [comment]
Thank you @bsardo . Quick question, when and what version will Prebid Server with our adapter be available for publishers?

**Pull request merged**
